File: writebook/app/models/p2u/compiler.rb
module P2u
  # Parses, validates and plans a manifest. Compilation is pure unless
  # `render: true` is passed, in which case diagrams and equations are built
  # through RenderedAsset and their stable URLs are returned in the plan.
  class Compiler
    Result = Struct.new(:manifest, :diagnostics, :plan, :valid, keyword_init: true) do
      def valid? = !!valid
      def errors = diagnostics.select(&:error?)
      def warnings = diagnostics.select(&:warning?)

      def to_h
        {
          valid: valid?,
          version: P2u::VERSION,
          manifest: manifest&.to_h,
          diagnostics: diagnostics.map(&:to_h),
          plan: plan
        }.compact
      end
    end

    def self.compile(source, format: nil, render: false)
      manifest = source.is_a?(Manifest) ? source : Parser.parse(source, format: format)
      new(manifest).compile(render: render)
    rescue ParseError => e
      failure = Diagnostic.new(severity: "error", code: "parse_error", message: e.message)
      Result.new(manifest: nil, diagnostics: [ failure ], plan: [], valid: false)
    end

    def initialize(manifest)
      @manifest = manifest
    end

    def compile(render: false)
      diagnostics = Validator.new(@manifest).validate
      plan = []

      if diagnostics.none?(&:error?)
        plan = asset_requests.map do |request|
          compile_asset(request, render: render, diagnostics: diagnostics)
        end
      end

      Result.new(
        manifest: @manifest,
        diagnostics: diagnostics,
        plan: plan,
        valid: diagnostics.none?(&:error?)
      )
    end

    private
      # Every asset-producing block in the deck, in slide order.
      def asset_requests
        @manifest.slides.flat_map do |slide|
          Array(slide["blocks"]).filter_map do |block|
            next unless block.is_a?(Hash)

            case block["kind"].to_s
            when "diagram" then diagram_request(slide, block)
            when "equation" then { slide: slide["id"], kind: "latex", source: block["source"].to_s, options: {} }
            end
          end
        end
      end

      def diagram_request(slide, block)
        engine = (block["lang"].presence || block["engine"].presence || "d2").to_s
        return { slide: slide["id"], kind: "typst", source: block["source"].to_s, options: {} } if %w[typst typ].include?(engine)
        return unless %w[d2 d2-sketch tala].include?(engine)

        options = { layout: "tala", theme: 301, sketch: engine == "d2-sketch" }
        options.merge!(P2u.deep_stringify(block["options"]).symbolize_keys) if block["options"].is_a?(Hash)

        { slide: slide["id"], kind: "d2", source: block["source"].to_s, options: options }
      end

      def compile_asset(request, render:, diagnostics:)
        digest = RenderedAsset.digest_for(
          kind: request[:kind], source: request[:source], options: request[:options]
        )

        entry = {
          slide: request[:slide],
          kind: request[:kind],
          digest: digest,
          url: "/rendered/#{digest}.svg",
          rendered: false
        }

        if render
          begin
            entry[:url] = RenderedAsset.store(
              kind: request[:kind], source: request[:source], options: request[:options]
            )
            entry[:rendered] = true
          rescue StandardError => e
            diagnostics << Diagnostic.new(
              severity: "error", code: "render_failed", slide: request[:slide],
              message: "#{request[:kind]} render failed: #{e.message}"
            )
          end
        end

        entry
      end
  end
end

File: cloudflare/src/compiler/compiler.ts
// Parses, validates and plans a manifest. Compilation is pure unless
// `render: true` is passed, in which case diagrams and equations are built and
// their stable URLs are returned in the plan. Port of p2u/compiler.rb.

import { Manifest } from "./manifest";
import type { Slide } from "./manifest";
import { Parser } from "./parser";
import { Validator } from "./validator";
import { Diagnostic, ParseError } from "./diagnostic";
import { presence } from "./util";
import { digestFor } from "../render/digest";
import type { AssetEntry, AssetKind, AssetRequest } from "../render/digest";

export interface CompileResult {
  manifest: Manifest | null;
  diagnostics: Diagnostic[];
  plan: AssetEntry[];
  valid: boolean;
}

export interface CompileOptions {
  format?: string;
  render?: boolean;
  /** Renders and stores an asset, returning its stable URL. Required when render is true. */
  store?: (request: AssetRequest) => Promise<string>;
}

export class Compiler {
  private readonly manifest: Manifest;

  constructor(manifest: Manifest) {
    this.manifest = manifest;
  }

  static async compile(source: unknown, options: CompileOptions = {}): Promise<CompileResult> {
    let manifest: Manifest;
    try {
      manifest = source instanceof Manifest ? source : Parser.parse(source, options.format);
    } catch (error) {
      const diagnostic =
        error instanceof ParseError
          ? new Diagnostic("error", "parse_error", error.message)
          : new Diagnostic("error", "parse_error", (error as Error).message);
      return { manifest: null, diagnostics: [diagnostic], plan: [], valid: false };
    }

    return new Compiler(manifest).compile(options);
  }

  async compile(options: CompileOptions = {}): Promise<CompileResult> {
    const diagnostics = Validator.validate(this.manifest);
    let plan: AssetEntry[] = [];

    if (!diagnostics.some((diagnostic) => diagnostic.isError)) {
      const requests = assetRequests(this.manifest);
      plan = [];
      for (const request of requests) {
        plan.push(await this.compileAsset(request, options, diagnostics));
      }
    }

    return {
      manifest: this.manifest,
      diagnostics,
      plan,
      valid: !diagnostics.some((diagnostic) => diagnostic.isError),
    };
  }

  private async compileAsset(
    request: AssetRequest,
    options: CompileOptions,
    diagnostics: Diagnostic[],
  ): Promise<AssetEntry> {
    const digest = await digestFor(request);
    const entry: AssetEntry = {
      slide: request.slide,
      kind: request.kind,
      digest,
      url: `/rendered/${digest}.svg`,
      rendered: false,
    };

    if (options.render) {
      if (!options.store) throw new Error("compile(render: true) requires a store callback");
      try {
        entry.url = await options.store(request);
        entry.rendered = true;
      } catch (error) {
        diagnostics.push(
          new Diagnostic("error", "render_failed", `${request.kind} render failed: ${(error as Error).message}`, {
            slide: request.slide,
          }),
        );
      }
    }

    return entry;
  }
}

/** Every asset-producing block in the deck, in slide order. */
export function assetRequests(manifest: Manifest): AssetRequest[] {
  const requests: AssetRequest[] = [];

  for (const slide of manifest.slides) {
    for (const block of slide.blocks ?? []) {
      if (!block || typeof block !== "object") continue;

      const kind = String((block as any).kind ?? "");
      if (kind === "diagram") {
        const request = diagramRequest(slide, block as Record<string, unknown>);
        if (request) requests.push(request);
      } else if (kind === "equation") {
        requests.push({ slide: slide.id, kind: "latex", source: String((block as any).source ?? ""), options: {} });
      }
    }
  }

  return requests;
}

function diagramRequest(slide: Slide, block: Record<string, unknown>): AssetRequest | null {
  const engine = String(presence(block.lang) ?? presence(block.engine) ?? "d2");
  if (engine === "typst" || engine === "typ") {
    return { slide: slide.id, kind: "typst", source: String(block.source ?? ""), options: {} };
  }
  if (!["d2", "d2-sketch", "tala"].includes(engine)) return null;

  const options: Record<string, unknown> = { layout: "tala", theme: 301, sketch: engine === "d2-sketch" };
  if (block.options && typeof block.options === "object" && !Array.isArray(block.options)) {
    Object.assign(options, block.options);
  }

  return { slide: slide.id, kind: "d2", source: String(block.source ?? ""), options };
}

export type { AssetEntry, AssetKind, AssetRequest };